crates/nu-command/src/filesystem/mv.rs

```diff
@@ -0,0 +1,141 @@
+use std::env::current_dir;
+use std::path::{Path, PathBuf};
+
+use nu_engine::CallExt;
+use nu_protocol::ast::Call;
+use nu_protocol::engine::{Command, EvaluationContext};
+use nu_protocol::{ShellError, Signature, SyntaxShape, Value};
+
+pub struct Mv;
+
+impl Command for Mv {
+    fn name(&self) -> &str {
+        "mv"
+    }
+
+    fn usage(&self) -> &str {
+        "Move files or directories."
+    }
+
+    fn signature(&self) -> nu_protocol::Signature {
+        Signature::build("mv")
+            .required(
+                "source",
+                SyntaxShape::GlobPattern,
+                "the location to move files/directories from",
+            )
+            .required(
+                "destination",
+                SyntaxShape::FilePath,
+                "the location to move files/directories to",
+            )
+    }
+
+    fn run(
+        &self,
+        context: &EvaluationContext,
+        call: &Call,
+        _input: Value,
+    ) -> Result<nu_protocol::Value, nu_protocol::ShellError> {
+        // TODO: handle invalid directory or insufficient permissions
+        let source: String = call.req(context, 0)?;
+        let destination: String = call.req(context, 1)?;
+
+        let path: PathBuf = current_dir().unwrap();
+        let source = path.join(source.as_str());
+        let destination = path.join(destination.as_str());
+
+        let mut sources =
+            glob::glob(&source.to_string_lossy()).map_or_else(|_| Vec::new(), Iterator::collect);
+
+        if sources.is_empty() {
+            return Err(ShellError::InternalError(format!(
+                "source \"{:?}\" does not exist",
+                source
+            )));
+        }
+
+        if (destination.exists() && !destination.is_dir() && sources.len() > 1)
+            || (!destination.exists() && sources.len() > 1)
+        {
+            return Err(ShellError::InternalError(
+                "can only move multiple sources if destination is a directory".to_string(),
+            ));
+        }
+
+        let some_if_source_is_destination = sources
+            .iter()
+            .find(|f| matches!(f, Ok(f) if destination.starts_with(f)));
+        if destination.exists() && destination.is_dir() && sources.len() == 1 {
+            if let Some(Ok(filename)) = some_if_source_is_destination {
+                return Err(ShellError::InternalError(format!(
```

Comment:
Nice!

My one suggestion is that instead of using ShellError::InternalError, can you add cases to ShellError for these different errors?

If you can't see how to do it, let me know and I can make a few suggestions.

Reply:
Awesome I was wondering about that.  I'll poke around implementations in other commands and get something together.  Will definitely reach out if I need help, thanks.